### DISCUSSION TITLE
i get a status_code 400 from valhalla while running meili functionality.

### DISCUSSION THREAD

**Discussion by sayaligo** [valhalla/valhalla#4619] [Q&A]
here is my full code - 

import os
import pandas as pd
import numpy as np
import requests
import json
from tqdm.notebook import tqdm

# Specify the file path to your CSV file
file_path = "Porto City.csv"

# Specify the columns you want to read (latitude, longitude, and time)
columns_to_read = ["latitude", "longitude", "pdatetime"]

# Read the CSV file into a DataFrame, reading only the specified columns
df_trip = pd.read_csv(file_path, usecols=columns_to_read)

# Display the DataFrame to verify the data
print(df_trip)

# Getting our Pandas DataFrame into a JSON
# We need that because that is a Meili's request requirement
meili_coordinates = df_trip.to_json(orient='records')

# Providing needed data for the body of Meili's request
meili_head = '{"shape":'
# Those are parameters that you can change according to the Meili's documentation
meili_tail = ""","search_radius": 150, "shape_match":"map_snap", "costing":"auto", "format":"osrm"}"""

# Combining all the string into a single request
meili_request_body = meili_head + meili_coordinates + meili_tail

# The URL of our local Valhalla server
# "trace_route" means that we're sening a request to Meili
url = "http://localhost:8002/trace_route"

# Providing headers to the request
headers = {'Content-type': 'application/json'}

# We need to send our JSON as a string
data = str(meili_request_body)

# Sending a request
r = requests.post(url, data=data, headers=headers)

if r.status_code == 200:
    
        response_text = json.loads(r.text)

        resp = str(response_text['tracepoints'])

        resp = resp.replace("'waypoint_index': None", "'waypoint_index': '#'")
        resp = resp.replace("None", "{'matchings_index': '#', 'name': '', 'waypoint_index': '#', 'alternatives_count': 0, 'distance': 0, 'location': [0.0, 0.0]}")

        resp = resp.replace("'", '"')

        resp = json.dumps(resp)
        resp = json.loads(resp)
        
        df_response = pd.read_json(resp)
        df_response = df_response[['name', 'distance', 'location']]
                
        df_trip_optimized = pd.merge(df_trip_for_meili, df_response, left_index=True, right_index=True)

print(r.status_code)
#print(r.text)

this r.status_code returns 400. what is this error ? AlSO, i have created a custom_files folder in c and linked it with my valhalla docker image. valhalls is running on port 8002. custom_files folder contains map of portugal city from geofabrik download server as i am passing coordinates of porto city. what am i doing wrong ? I am new to valhalla. Any urgent help would be appreciated.

**Comment by sayaligo** [comment]
I removed format = osrm and now it gives me error code = 134, failed to parse shape, I am passing long, lat and time in shape

**Reply by kevinkreiser** [reply]
you'll have to give us the actual request or an example that looks like it so we can tell you what is wrong with it. or just head over to the docs for that service and have a look at what the input should look like. theres an example at the bottom of this section: https://valhalla.github.io/valhalla/api/map-matching/api-reference/#get-better-results

**Reply by sayaligo** [reply]
Hi Kevin,I am following [https://tow](https://towardsdatascience.com/map-matching-done-right-using-valhallas-meili-f635ebd17053)

**Comment by nilsnolde** [comment]
We need an actual request, not the code to generate the request. Smth which we can curl.

**Reply by sayaligo** [reply]
Hi Nils, I am following https://towardsdatascience.com/map-matching-done-right-using-valhallas-meili-f635ebd17053.

**Reply by nilsnolde** [reply]
I’m not going to read that though. Like I said, and Kevin, we need the payload of an actual HTTP POST request to be able to support..

**Reply by kevinkreiser** [reply]
i yes i remember that article now, the code you pasted rang a bell but i just couldnt remember it. anyway you should be able to step through it with your data and at least see the request that python makes. then you can post it here or compare it with the docs

**Reply by sayaligo** [reply]
I am confused how should i pass the meili coordinates, its head and tail. I am trying a sample data just to see how the meili functionality works. I am passing a csv file with latitude and longitude. do i need to pass anything else ? also, whats the process of keeping the .osm.pbf files in the folder which can be accessed by docker ? for example if i am passing coordinates of portugal then i should be keeping .osm.pbf file of portugal. right ? Please correct my understanding if wrong.

**Reply by nilsnolde** [reply]
you'll need to familiarize yourself with this software, no way around that, e.g. https://valhalla.github.io/valhalla/building/#running-valhalla-server-on-unix. a pbf needs to be turned into a valhalla graph.

**Comment by sayaligo** [comment]
I was able to make the code running for meili but seems the map matching is not working. I am sharing the images before (all pink dots)and after map matching for andorra.
![IMG_6673](https://github.com/valhalla/valhalla/assets/108901954/131acd12-6c9b-464d-b6c6-9152e80b1512)
![IMG_6672](https://github.com/valhalla/valhalla/assets/108901954/c32c2393-9cb2-4ab2-ae30-e9b53d2b0b81)
(url)

**Reply by sayaligo** [reply]
Also, i have installed valhalla using docker on windows. How can i add tiles for different pbf files. I tried adding new pnf file for another country in custom_files folder and restarted the docker but that did not built new tile file.how can i use option use_tiles_ignore_pbf= False on windows ? Whats the complete command ?
